py/src/nupic/external/hexy.py

```diff
@@ -0,0 +1,235 @@
+# ------------------------------------------------------------------------------
+# Copyright (c) 2017 Norbu Tsering.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+# Source:  https://github.com/RedFT/Hexy
```

Comment:
>  py/src/nupic/external/hexy.py 

nit, py/external/hexy ?

Reply:
If hexy was in `py/external` then it would need to be installed separately.
Since hexy is in `py/src/nupic/external` it is still part of the nupic distribution.
Currently the grid-cell code imports hexy as `import nupic.external.hexy`

Reply:
> Currently the grid-cell code imports hexy as import `nupic.external.hexy`

I assume that is what we want. correct?  So CMake does not need to do anything.

Reply:
>  CMake does not need to do anything. 

I think the point is if the install step should put python somewhere in the install directory (Release/), or if py code stays within the repo (in py/), right?

Reply:
Yes, the install step needs to copy the python code into the release dir, which currently does notappear to work.

Reply:
I guess I do not understand.  Why does it need to be in the release dir?  When we get all of the code from nupic.py does it also need to be copied to the install dir?

Reply:
> When we get all of the code from nupic.py does it also need to be copied to the install dir?

Yes.

Reply:
Here are the steps that need to happen for python installations to work:

1) Several peices of code need to end up in the python package:
    - The compiled bindings, from pybind11
    - `bindings/py/packaging/` which contains code needed to import the bindings & make them usable.
    - The source code for the grid cell encoder
    - Eventually, the contents of `nupic.py`

2) All of these peices of code are gathered in a single location: `build/Release/distr/src/nupic/`. 

3) The programs `setup.py` & `bindings/py/packaging/setup.py` take the contents of the  `distr` directory and package them into a format which python can install.

4) The command: `$ python3 setup.py install` copies the nupic package to a location on the users filesystem where python knows to look for it.

5) The user's program writes the python statement `import nupic`.  Python searches  `$PATH` and `$PYTHONPATH`  for a file which starts with "nupic" and has an acceptable file extension such as `.py` or `.egg`.  Python does not know to search through this git repo.


Somehow, the grid cell encoder is not being included in the final nupic distribution.  I think that step 2 is not working.

Reply:
> If hexy was in py/external then it would need to be installed separately.
Since hexy is in py/src/nupic/external it is still part of the nupic distribution. 

I have an update about hexy, 
the author is amazingly responsive, so we can now `pip install hexy` ! 
I think it's the proper way to go. 

It has some disadvantages: 
- updates numpy to 1.15 (should not be a problem?) 
- installs all its dependencies (including visualization stuff we don't need) 

Mitigations: 
- setup.py / pip has a nice option of splitting dependencies to required/extras, which I've submitted upstream : https://github.com/RedFT/Hexy/issues/4 
- or we could just uninstall the unneeded stuff after hexy install
- or actually keep at there and use it? (the demo example for Coordinate, RDSE? encoder uses matplotlib (if installed), could you use hexy instead for the visualization?)

Reply:
Regarding the problem with installation of python files

> got CMake to work on my machine, but it does not work in CI... Can you take a look at the file: py/CMakeLists.txt and see if you can tell what the problem is? 

why are we dealing with CMake there? I think we should keep the separation:
- cmake installs c++ 
- python setup.py installs python (and its bindings..)
  - so these files (encoder in py) should be included from setup.py

Reply:
> pip install hexy

That is cool.

But for the rest of the .py code....what you are saying is you think that all of the nupic.py code should end up in a .egg file so that the end user installs it all using `pip install nupic` from PyPI rather than downloading the repository.  Are we really ready to do that yet?

Reply:
> py code should end up in a .egg file so that the end user installs it all using pip install nupic from PyPI rather than downloading the repository. Are we really ready to do that yet? 

not fully, but yes.. we want to get to state that python code from this repo is installed by 
`python setup.py install`, so the user doesn't have to be "inside the repo folder" to run py scripts. 
Otherwise the `from nupic.bindings.encoders.xxx` would fail

Reply:
Ok, I will start looking into this.  The goal being to install all of nupic.py code.

Reply:
Thanks!

> The goal being to install all of nupic.py code.

Eventually, but now only what is in `/py/`